apiserver/objects_test.go

```diff
@@ -219,11 +326,276 @@ func (s *putCharmObjectSuite) TestUploadBumpsRevision(c *gc.C) {
 	c.Assert(sch.BundleSha256(), gc.Not(gc.Equals), "")
 }
 
+func (s *putCharmObjectSuite) TestUploadVersion(c *gc.C) {
+	expectedVersion := "dummy-146-g725cfd3-dirty"
+
+	// Add the dummy charm with version "juju-2.4-beta3-146-g725cfd3-dirty".
+	pathToArchive := testcharms.Repo.CharmArchivePath(c.MkDir(), "dummy")
+	err := testcharms.InjectFilesToCharmArchive(pathToArchive, map[string]string{
+		"version": expectedVersion,
+	})
+	c.Assert(err, gc.IsNil)
+	ch, err := charm.ReadCharmArchive(pathToArchive)
+	c.Assert(err, gc.IsNil)
+
+	f, err := os.Open(ch.Path)
+	c.Assert(err, jc.ErrorIsNil)
+	defer f.Close()
+	resp := s.uploadRequest(c, s.objectsCharmsURI("dummy-"+getCharmHash(c, f)), "application/zip", "local:dummy", f)
+
+	expectedURL := "local:dummy-1"
+	s.assertUploadResponse(c, resp, expectedURL)
+	sch, err := s.ControllerModel(c).State().Charm(expectedURL)
+	c.Assert(err, jc.ErrorIsNil)
+
+	version := sch.Version()
+	c.Assert(version, gc.Equals, expectedVersion)
+}
+
+func (s *putCharmObjectSuite) TestUploadRespectsLocalRevision(c *gc.C) {
+	// Make a dummy charm dir with revision 123.
+	dir := testcharms.Repo.ClonedDir(c.MkDir(), "dummy")
+	dir.SetDiskRevision(123)
+	// Now bundle the dir.
+	tempFile, err := os.CreateTemp("", "charm")
+	c.Assert(err, jc.ErrorIsNil)
+	defer tempFile.Close()
+	defer os.Remove(tempFile.Name())
+	err = dir.ArchiveTo(tempFile)
+	c.Assert(err, jc.ErrorIsNil)
+
+	expectedSHA256 := getCharmHash(c, tempFile)
+
+	// Now try uploading it and ensure the revision persists.
+	resp := s.uploadRequest(c, s.objectsCharmsURI("dummy-"+expectedSHA256), "application/zip", "local:dummy", tempFile)
+	expectedURL := "local:dummy-123"
+	s.assertUploadResponse(c, resp, expectedURL)
+	sch, err := s.ControllerModel(c).State().Charm(expectedURL)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(sch.URL(), gc.Equals, expectedURL)
+	c.Assert(sch.Revision(), gc.Equals, 123)
+	c.Assert(sch.IsUploaded(), jc.IsTrue)
+	c.Assert(sch.BundleSha256()[0:7], gc.Equals, expectedSHA256)
+
+	store := testing.NewObjectStore(c, s.ControllerModelUUID(), s.ControllerModel(c).State())
+	reader, _, err := store.Get(context.Background(), sch.StoragePath())
+	c.Assert(err, jc.ErrorIsNil)
+	defer reader.Close()
+	downloadedSHA256, _, err := utils.ReadSHA256(reader)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(downloadedSHA256[0:7], gc.Equals, expectedSHA256)
+}
+
+func (s *putCharmObjectSuite) TestUploadWithMultiSeriesCharm(c *gc.C) {
```

Comment:
Is this test a hold over from juju 1.x when multi series charms were introduced. Is it still really needed and different from the others? Or is it other tests that should be removed, as technically all charms are multi series now.

Reply:
Yes good point. This test looks like a pretty exact duplicate of a previous test

I'll remove